### PR TITLE
=htp #16962 docs for creating custom directives

### PR DESCRIPTION
**Pull request by johanandren**
Mostly ported from the spray docs

Refs #16962

### CODE REVIEW

**Comment by johanandren** [comment]
Haven't added anything to the Java side of the docs though

File: akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala [hunk 1]
```diff
@@ -68,6 +69,7 @@ trait HeaderDirectives {
     headerValue(f).map(Some(_): Option[T]).recoverPF {
       case Nil ⇒ provide(None)
     }
+  //#
```
Comment (ktoso) [inline akka-http/src/main/scala/akka/http/scaladsl/server/directives/HeaderDirectives.scala]: AFAIR includecode2 can work with `#optionalHeaderValue` for those. But maybe OK to include with docs in any case :)

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst [hunk 2]
```diff
@@ -3,4 +3,194 @@
...
+custom directives at differing levels of abstraction.
+
+There are essentially three ways of creating custom directives:
+
+1. By introducing new “labels” for configurations of existing directives
+2. By transforming existing directives
+3. By writing a directive “from scratch”
+
+Configuration Labeling
+______________________
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: I thought most docs use `-----` here hm...

File: akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/CustomDirectivesExamplesSpec.scala [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+  "labeling" in {
+    val getOrPut = get | put
+
+    // tests:
+    val route = getOrPut { complete("ok") }
+
+    Get("/") ~> route ~> check {
+      responseAs[String] shouldEqual "ok"
+    }
```
Comment (ktoso) [inline akka-docs-dev/rst/scala/code/docs/http/scaladsl/server/directives/CustomDirectivesExamplesSpec.scala]: thanks for keeping the style including the `// tests:` bit :)

**Comment by ktoso** [comment]
LGTM, looks good.

**Comment by ktoso** [comment]
FYI, for Java it's slightly different, people can implement `RequestVal`s but not really advanced directives.

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst (hunk 2 above)
Comment (johanandren) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: There was two levels of small headers in the original, thats why I used them.

Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: I see, as long as it looks good I'm happy :)

**Comment by johanandren** [comment]
Should I add something about the Java side of things as well in this PR?

**Comment by ktoso** [comment]
Can be a separate PR if you'd prefer that (can I count on you doing the Java side for this? :-))

**Comment by johanandren** [comment]
I don't really have an opinion about it. I'll look at what needs to be done and include in this PR

**Comment by ktoso** [comment]
Cool, thanks! The most they can do is provide some custom request val, not much more I guess. // cc @jrudolph

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst [hunk 4]
```diff
@@ -3,4 +3,194 @@
...
+
+A common pattern that relies on flatMap is to first extract a value
+from the RequestContext with the extract directive and then flatMap with
+some kind of filtering logic. For example, this is the implementation
+of the method directive:
+
+.. includecode2:: ../../../../../../akka-http/src/main/scala/akka/http/scaladsl/server/directives/MethodDirectives.scala
+   :snippet: method
+
+The explicit type parameter ``[Unit]`` on the flatMap i`s needed in this case
```
Comment (rkuhn) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: there’s a `\`` too many here

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst (hunk 2 above)
Comment (rkuhn) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: so far we used `---` for the second and `^^^` for the third level (IIRC)

File: akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst (hunk 4 above)
Comment (ktoso) [inline akka-docs-dev/rst/scala/http/routing-dsl/directives/custom-directives.rst]: I fixed that and will merge

**Comment by ktoso** [comment]
I'll merge this and we can go ahead with other thingies soon

**Comment by ktoso** [comment]
This is merged now (though manually, since I fixed a typo pointed out above): 39034f3d687d00d2e206a3835909d8a4a77006aa

**Comment by ktoso** [comment]
Thanks @johanandren !

**Pull request closed without merging**